fix(zone.js): remove global declaration

JiaLiPassion opened the pull request:
Close #37531

Remove `global` declaration in zone.ts to avoid compile error when
upgrade to `@types/node`.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: 37531


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

mhevery commented:
[presubmit](http://test/OCL:321604106:BASE:321604156:1594922892130:a09509ed)

mhevery commented:
CARETAKER: Needs cl/321604430

mhevery commented:
This is failing with 
```
third_party/javascript/angular2/rc/packages/zone.js/lib/zone.ts:1437:86 - error TS2304: Cannot find name 'global'.

1437 })(typeof window !== 'undefined' && window || typeof self !== 'undefined' && self || global);
```

I think adding 
```
(typeof window !== 'undefined' && window || typeof self !== 'undefined' && self || typeof global !== 'undefined' && global);
```

solves the problem.

JiaLiPassion commented:
@mhevery , got it, thanks, I just updated the `zone.ts`.

mhevery commented:
[presubmit](http://test/OCL:321604106:BASE:323396881:1595872926994:2beef057)

mhevery commented:
Argh! Seems like that did not help. :-(

```
third_party/javascript/angular2/rc/packages/zone.js/lib/zone.ts:1438:11 - error TS2304: Cannot find name 'global'.

1438    typeof global !== 'undefined' && global);
               ~~~~~~
third_party/javascript/angular2/rc/packages/zone.js/lib/zone.ts:1438:37 - error TS2304: Cannot find name 'global'.

1438    typeof global !== 'undefined' && global);
                                         ~~~~~~
```

JiaLiPassion commented:
@mhevery, The motivation is the issue #37531, 
```
In DefinitelyTyped/DefinitelyTyped#44700, the type definition of the global variable was updated to NodeJS.Global & typeof globalThis. Which makes the Zone.js package not compatible with @types/node 14.0.5 or up.
```

I will update the `global` type definition to match the new `nodejs` definition.

JiaLiPassion commented:
@mhevery , for now I updated the global declaration to `any` just like https://github.com/angular/angular/blob/master/packages/core/src/util/global.ts#L14. 
After we updated `@types/node` to `v14+`, I will update the `global` type again.

mhevery commented:
Hmm, it seems like the build system does not like that.

dag03tsc commented:
The `@types/node` module loads before the `zone.js` one and it declares the global variable so its type can't be changed later.

Therefore, I think the solution could be, in fact, to exactly mirror the global declaration (Or maybe remove it instead?).

I've bypassed this since the bug appeared by applying the next patch after `npm install`.

> --- node_modules/zone.js/dist/zone.js.d.ts
> +++ node_modules/zone.js/dist/zone.js.d.ts
> @@ -600,34 +600,54 @@
> -declare var global: NodeJS.Global;
> +declare var global: NodeJS.Global & typeof globalThis;
> 
> --- node_modules/zone.js/lib/zone.d.ts
> +++ node_modules/zone.js/lib/zone.d.ts
> @@ -600,34 +600,54 @@
> -declare var global: NodeJS.Global;
> +declare var global: NodeJS.Global & typeof globalThis;

It doesn't generate any side effects and everything seems alright so far.

I guess the global type should be declared as `NodeJS.Global & typeof globalThis` instead of `any`.

Maybe removing it could solve it too, just by letting the `@types/node` handle it all, but I'm not sure if it'd carry any unwanted side effects due to the missing declaration if, somehow, `@types/node` isn't available, so the first option seems safer.

kenaidri commented on packages/zone.js/lib/zone.ts:
```diff
@@ -678,7 +678,7 @@ type AmbientZoneDelegate = ZoneDelegate;
 // This code should run in a Browser, so we don't want to include the whole node.d.ts
 // typings for this compilation unit.
 // We'll just fake the global "global" var for now.
-declare var global: NodeJS.Global;
```
Thanks

dag03tsc replied:
Oh! I see! I didn't knew about the suggestion feature. Thanks @kenaidri!

dag03tsc commented on packages/zone.js/lib/zone.ts:
```diff
@@ -678,7 +678,7 @@ type AmbientZoneDelegate = ZoneDelegate;
 // typings for this compilation unit.
 // We'll just fake the global "global" var for now.
-declare var global: NodeJS.Global;
+declare var global: any;
```
```suggestion
declare var global: NodeJS.Global & typeof globalThis;
```
Make sure that the global type match the one provided by the current `@types/node` module.

dag03tsc commented on packages/zone.js/package.json:
```diff
@@ -19,7 +19,7 @@
   },
   "devDependencies": {
-    "@types/node": "^10.9.4",
+    "@types/node": "^14.0.5",
```
```suggestion
    "@types/node": "^14.0.27",
```
Upgrade `@types/node` to last version.

JiaLiPassion replied:
Yeah, we have to wait for node in `angular` repo update to `14.0.27+` and continue on this PR later.

dag03tsc replied:
I see. Nevermind!

dag03tsc commented on packages/zone.js/test/common/zone.spec.ts:
```diff
@@ -343,7 +343,7 @@ describe('Zone', function() {
         const ZoneAwarePromise = global.Promise;
         try {
-          global.Promise = WrongPromise;
+          (global as any).Promise = WrongPromise;
```
```suggestion
          global.Promise = WrongPromise;
```
I'm not actually sure about this one. How do you think it should be done?
 - [X]  Leave global uncasted as it was before.
 - [ ]  Replace the `any` global cast with a `(NodeJS.Global & typeof globalThis)` one.
 - [ ]  Don't change anything at all.

dag03tsc reviewed (changes requested):
Upgrade `@types/node` from `v14.0.5` to `v14.0.27`.
Mirror the global declaration provided by the `@types/node` module ([breaking change](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44700)).

Edit: I apologize for sending 4 reviews in a row. It's been an error due to my lack of experience with the suggestion feature! I've learnt the lesson and next times I'll start a review instead of adding single comments!

JiaLiPassion commented:
@dag03tsc , no problem, thank you for the review, `nodejs` update may not happen right now, will continue working on this PR after that.

josephperrott reviewed (approved):
LGTM

Reviewed-for: dev-infra

JiaLiPassion committed:
fix(zone.js): remove global declaration

Close #37531

Remove `global` declaration in `zone.ts` to avoid compile error when
upgrade to `@types/node` v12.12.68. Since the new type of global become
`NodeJS.global & typeof globalThis` and not compatible with `zone.ts` declaration.

josephperrott commented:
Caretaker Note: This will require cl/338344996 to land before it can be merged.

AndrewKushnir commented:
@josephperrott @JiaLiPassion based on the "target: major" label and the current state of release train, this PR will be included into v12 (and will not land in upcoming v11). The target was set a couple months ago, so I believe the intention was to land it in v11, so should we change the label to "target: rc" to reflect that? Please confirm.

JiaLiPassion commented:
@AndrewKushnir , yes, this one is targeting `v11`, I have updated the label to `target: rc`.
